File: 1289.minimum-falling-path-sum-ii.js
/*
 * @lc app=leetcode id=1289 lang=javascript
 *
 * [1289] Minimum Falling Path Sum II
 */

// @lc code=start
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minFallingPathSum = function (grid)
{
    /**
     * m = grid.length
     * n = grid[0].length
     * 
     * dp[i][j] 到第 i 层时，从 j 开始下降，最小下降成本是多少？
     * 
     * base case
     * 
     * dp[m-1][j] = grid[m-1][j]
     * 
     * dp[i][j] = grid[i][j] + min(dp[i+1][k]) k!==j
     * 
     * 内存压缩 
     * dp = dp[i]
     * prevDp = dp[i+1]
     * 
     * base case
     * 
     * prevDp[j] = grid[m-1][j]
     * 
     * dp[j] = grid[i][j] + min(prevDp[k]) k!==j
     */

    const m = grid.length;
    const n = grid[0].length;

    let prevDp = grid[m - 1];
    let dp = new Array(m);

    for (let i = m - 2; i >= 0; i--)
    {
        for (let j = 0; j < n; j++)
        {
            let minRestSum = Infinity;
            for (let k = 0; k < n; k++)
            {
                if (k !== j)
                {
                    minRestSum = Math.min(minRestSum, prevDp[k]);
                }
                dp[j] = grid[i][j] + minRestSum;
            }
        }
        prevDp = dp;
        dp = new Array(m);
    }

    return Math.min(...prevDp);
};
// @lc code=end

File: Daily/1289.minimum-falling-path-sum-ii.cpp
/*
 * @lc app=leetcode id=1289 lang=cpp
 *
 * [1289] Minimum Falling Path Sum II
 */
#include <climits>
#include <cmath>
#include <queue>
#include <utility>
#include <vector>

using std::pair;
using std::priority_queue;
using std::vector;

// @lc code=start
class Solution {
 public:
  int minFallingPathSum(vector<vector<int>>& grid) {
    const int N = grid.size();
    vector<vector<int>> dp(2);
    for (int i = 0; i < 2; i++) {
      dp[i].resize(N);
    }
    dp[0] = grid[0];

    for (int i = 1; i < N; i++) {
      auto lastDpRowMinTwoCols = findMinTwoNumsIndexes(dp[(i - 1) % 2]);
      for (int j = 0; j < N; j++) {
        dp[i % 2][j] = dp[(i - 1) % 2][lastDpRowMinTwoCols.first == j
                                           ? lastDpRowMinTwoCols.second
                                           : lastDpRowMinTwoCols.first] +
                       grid[i][j];
      }
    }

    return *std::min_element(dp[(N - 1) % 2].cbegin(), dp[(N - 1) % 2].cend());
  }

 private:
  pair<int, int> findMinTwoNumsIndexes(const vector<int>& nums) {
    int minIndex = -1;
    int secondMinIndex = -1;

    for (int i = 0; i < nums.size(); i++) {
      if (minIndex == -1 || nums[i] < nums[minIndex]) {
        secondMinIndex = minIndex;
        minIndex = i;
      } else if (secondMinIndex == -1 || nums[i] < nums[secondMinIndex]) {
        secondMinIndex = i;
      }
    }

    return {minIndex, secondMinIndex};
  };
};
// @lc code=end

int main() {
  Solution sol;
  vector<vector<int>> vec = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
  sol.minFallingPathSum(vec);
}